### PR TITLE
v9 availability rules

### PR DESCRIPTION
**Pull request by jannotti**
First cut at new group-level resource sharing specs.

### CODE REVIEW

File: dev/TEAL.md [hunk 1]
```diff
@@ -201,6 +201,40 @@ _available_.
...
+     1. `pay` - `txn.Sender`, `txn.Receiver`, and `txn.CloseRemainderTo`
+
+     1. `keyreg` - `txn.Sender`
+
+     1. `acfg` - `txn.Sender`, `txn.ConfigAsset`, and the
+        `txn.ConfigAsset` holding of `txn.Sender`.
+
+     1. `axfer` - `txn.Sender`, `txn.AssetSender`, `txnAssetCloseTo`,
+        `txn.XferAsset` and the `txn.XferAsset` holding of each of
+        those accounts.
```
Comment (nullun) [inline dev/TEAL.md:231]: Is `AssetReceiver` also available in an `axfer`?

Comment (jannotti) [reply dev/TEAL.md:231]: Yes, I will fix.

File: dev/TEAL.md [hunk 2]
```diff
@@ -201,6 +201,41 @@ _available_.
...
+   available.
+     
+ * Top-level transactions that are not application calls also make
+   resources available to group-level resource sharing. The following
+   resources are made available other transaction types.
+
+     1. `pay` - `txn.Sender`, `txn.Receiver`, and
+        `txn.CloseRemainderTo` (if set)
+
+     1. `keyreg` - `txn.Sender`
```
Comment (fabrice102) [inline dev/TEAL.md:225]: minor: replace 1,1,1,... by 1,2,3,4,...

File: dev/TEAL.md [hunk 3]
```diff
@@ -201,6 +201,41 @@ _available_.
...
+
+     1. `acfg` - `txn.Sender`, `txn.ConfigAsset`, and the
+        `txn.ConfigAsset` holding of `txn.Sender`.
+
+     1. `axfer` - `txn.Sender`, `txn.AssetReceiver`, `txn.AssetSender`
+        (if set), `txnAssetCloseTo` (if set), `txn.XferAsset` and the
+        `txn.XferAsset` holding of each of those accounts.
```
Comment (fabrice102) [inline dev/TEAL.md:232]: Including the sender account balances, even in case of a clawback?

File: dev/TEAL.md (hunk 2 above)
Comment (jannotti) [reply dev/TEAL.md:225]: I think that markdown causes the numbers to come out consecutively, but I might be mixing md up with some other format. I'll check.

File: dev/TEAL.md (hunk 3 above)
Comment (jannotti) [reply dev/TEAL.md:232]: Yes, I decided it was not a big deal to allow a little more access, as this is still much less I/O than an appl could generate. So I decided for simplicity of spec over precisely matching what is actually needed.

(As a further crazy example - axfer and pay of 0 does not actually read in the recipient, but I certainly don't want to spec that.)

File: dev/TEAL.md (hunk 2 above)
Comment (jannotti) [reply dev/TEAL.md:225]: Yes, repeated 1 is ok:
https://www.markdownguide.org/basic-syntax/#lists-1

File: dev/TEAL.md (hunk 3 above)
Comment (shaih) [reply dev/TEAL.md:232]: just curiosity, so you will fetch this information only if an application call needs it? Or always starting v9?

Comment (jannotti) [reply dev/TEAL.md:232]: The new changes will not cause it to be prefetched.  However, I think we _already_ prefetch this to speed up validation (not proposals, just validating proposals).  That prefetch code does not understand the subtle reasons why the holding might not always be needed. (I don't think.)

At any rate - no change in prefetching will happen due to this change.

Comment (jannotti) [reply dev/TEAL.md:232]: Actually, that code is fairly clever, so it *won't* be prefetched, before or now.
```
if !stxn.Txn.AssetSender.IsZero() {
	loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	loadAccountsAddResourceTask(&stxn.Txn.AssetSender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
} else {
	if stxn.Txn.AssetAmount == 0 && (stxn.Txn.AssetReceiver == stxn.Txn.Sender) { // opt in
		loadAccountsAddResourceTask(nil, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	}
	if stxn.Txn.AssetAmount != 0 { // zero transfer is noop
		loadAccountsAddResourceTask(&stxn.Txn.Sender, basics.CreatableIndex(stxn.Txn.XferAsset), basics.AssetCreatable, task, resourceTasks, queue)
	}
}
```

File: dev/TEAL.md [hunk 4]
```diff
@@ -201,6 +201,41 @@ _available_.
...
+   state is available by group-level resource sharing, the holding or
+   local state must be available in a top-level transaction without
+   considering group sharing. For example, if account A is made
+   available in one transaction, and asset X is made available in
+   another, group resource sharing does _not_ make A's X holding
+   available.
+     
+ * Top-level transactions that are not application calls also make
+   resources available to group-level resource sharing. The following
+   resources are made available other transaction types.
```
[the remark this answers is not in the document]
Comment (jannotti) [reply dev/TEAL.md:220]: will change

**Pull request merged**